packages/slate/src/controllers/editor.js

```diff
@@ -95,7 +95,7 @@ class Editor {
 
     // Apply the operation to the value.
     debug('apply', { operation })
-    this.value = operation.apply(value)
+    this.value = operation.apply(value).resolveRanges(this.controller)
```

Comment:
Is it OK for someone to create an editor with a `null` value? if so, this will break.

Reply:
I'm okay to have value always default to some empty version of itself.

Reply:
Since we don't have a concept of "default block," I don't think we can create a default empty value. I was more wondering whether we should expect people to pass in an empty value, or if passing in null was OK. Passing null doesn't _really_ work today, because eventually we'll try to call a method on it. But it works well enough to construct the editor and add one later, whereas with this, we expect it immediately.

I actually attached this comment to the wrong line. I meant this one: https://github.com/ianstormtaylor/slate/pull/2933/files#diff-a9ac6defda08b6f3742a41b794d18a3aL391

Reply:
Could we create a value that has an empty document? And just not have any blocks in it?

Reply:
You could, but selection would be invalid by definition, since there's no text node for it to go into. A document with direct text node children would have those children normalized out, because document can only have block node children. Either of those could be relaxed, but both would be a little bit of a weird scenario. I think I'd prefer the error in that case.

Reply:
Oh hey, #2991 would make that first one moot 😄

Reply:
Right now we can use the "unset" selection in the interim. But having it solved by #2991 would be awesome.

I'm pretty sure I'd like to have documents be able to contain no nodes. I use it in one of my situations right now, instead of having a strange "placeholder" node in there just to be able to create a value.

Reply:
Cool. I think it might make sense then to have `normalizeSelection` use Value's `mapRanges`, instead of partially reimplementing it. I think I ran into a problem with that before, but I've fixed enough problems since then that it's worth another try.